internal/xds/translator/translator.go

```diff
@@ -125,65 +126,150 @@ func (t *Translator) processHTTPListenerXdsTranslation(tCtx *types.ResourceVersi
 			}
 		}
 
-		// Allocate virtual host for this httpListener.
-		// 1:1 between IR HTTPListener and xDS VirtualHost
-		vHost := &routev3.VirtualHost{
-			Name:    httpListener.Name,
-			Domains: httpListener.Hostnames,
-		}
 		protocol := DefaultProtocol
 		if httpListener.IsHTTP2 {
 			protocol = HTTP2
 		}
 
-		// Check if an extension is loaded that wants to modify xDS Routes after they have been generated
+		routesByHostname := make(map[string][]*ir.HTTPRoute, len(httpListener.Routes))
 		for _, httpRoute := range httpListener.Routes {
-			// 1:1 between IR HTTPRoute and xDS config.route.v3.Route
-			xdsRoute := buildXdsRoute(httpRoute, xdsListener)
-
-			// Check if an extension want to modify the route we just generated
-			// If no extension exists (or it doesn't subscribe to this hook) then this is a quick no-op.
-			if err := processExtensionPostRouteHook(xdsRoute, vHost, httpRoute, t.ExtensionManager); err != nil {
-				return err
+			if len(httpRoute.Hostnames) == 0 {
+				// no hostnames => just use the listener
+				for _, hostname := range httpListener.Hostnames {
+					routesByHostname[hostname] = append(routesByHostname[hostname], httpRoute)
+				}
+			} else {
+				for _, hostname := range httpRoute.Hostnames {
+					routesByHostname[hostname] = append(routesByHostname[hostname], httpRoute)
+				}
 			}
+		}
+		hostnames := make([]string, 0, len(routesByHostname))
+		for hostname := range routesByHostname {
+			hostnames = append(hostnames, hostname)
+		}
+		// Sort according to rules in https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.HTTPRouteSpec
```

Comment:
all sorting should happen before hand
https://github.com/envoyproxy/gateway/blob/main/internal/gatewayapi/sort.go